xgcm/transform.py

```diff
@@ -18,8 +18,9 @@
     nopython=True,
 )
 def _interp_1d_linear(phi, theta, target_theta_levels, mask_edges, output):
-    # if last theta value is smaller than first, assume the profile is monotonically decreasing and flip
-    if theta[-1] < theta[0]:
+    # rough check if the data is decreasing with depth. If that is the case, flip.
+    theta_sign_test = theta[~np.isnan(theta)]
+    if theta_sign_test[-1] < theta_sign_test[0]:
```

Comment:
I don't understand how these checks can possibly be dask friendly. At this point, `theta` is still ND, no? Can you talk me through what you think is happening here?

Reply:
I might be understanding the way gufuncs work, wrongly. I thought that we can basically treat 'theta' as 1D at this point, since the function is mapped onto 1D columns of the data (I basically thought the numba wrapper implements a very efficient for loop for every 1D column). But it seems that is not the case?
Do you know how to make `all()` work in numba? That way we could use your above suggestion.

Reply:
Ah sorry, I was confused. Here we are inside the gufunc. You are not misunderstanding. Your code makes total sense.

Now my only question is: how expensive is this? What does it cost us to do this check all the time.

Reply:
This should be fine since this is the bit that is vectorized.

re: all in numba: use `np.all`: https://numba.readthedocs.io/en/stable/developer/autogen_numpy_listing.html#listings-for-numpy

Reply:
It's vectorized, but it still takes CPU time. How much? How much slower is the function with / without this check?

The reason I ask is that we might prefer to have a flag like `check=True` that we can turn on / off to bypass.

Reply:
I like that idea. I'd say we leave it on by default, and add a short section on how to tune performance (e.g. turn this check off if your data doesn't need to be flipped)?

Would be really nice to have #238 to be able to judge how this affects performance.

Reply:
We don't need #238 to judge how this affects performance.

We can just manually time something with / without it and report the answer right here in this thread.

Reply:
Will do that.

Reply:
OK I have come up with a simple synthetic test sample:
```python
import xarray as xr
import numpy as np
from xgcm import Grid

# create a dummy array
nx = 1000
ny = 1000
nz = 50
data = xr.DataArray(np.random.rand(nx, ny, nz), dims=['x','y','z'])

target = xr.DataArray(np.arange(nz), dims=['z']).expand_dims(x=nx, y=ny)
# add some noise to target
target = target + 0.2 * xr.DataArray(np.random.rand(nx, ny, nz), dims=['x','y','z'])

ds = xr.Dataset({'data':data, 'target':target})

grid = Grid(ds, coords={'Z':{'center':'z'}}, periodic=False)
```

I have quickly hacked together a bypass switch for the checks and timed the transform
```python
%%timeit 
grid.transform(ds.data, 'Z', np.arange(2,23,1.4), target_data=ds.target)
```
```
1.58 s ± 41.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```python
%%timeit 
grid.transform(ds.data, 'Z', np.arange(2,23,1.4), target_data=ds.target, bypass_checks=True)
```
```
1.22 s ± 82.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So there is a performance penalty, in this case ~30%. 

I think this is not negligible, so we should probably build this into the actual method? Id still say we should keep the checks on by default and add a 'performance tweaking' section to the docs.

Thoughts?

Reply:
Just a quick ping to @rabernat. Id go ahead and implement this next week, unless you think it isn't a good idea?